analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs

```diff
@@ -190,5 +190,102 @@ public void RecursivePattern_SetsNotNull_Nested()
+            var validator = SETestContext.CreateCS(code, ", object arg", setter).Validator;
+            validator.ValidateContainsOperation(OperationKind.DeclarationPattern);
+            validator.ValidateTag("Value", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("Value", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeFalse());     // ToDo: MMF-2563 should BeTrue() instead
+            validator.TagValues("End").Should().HaveCount(2).And.OnlyContain(x => x != null && x.HasConstraint(TestConstraint.First));  // 2x because value has different states
+        }
+
+        [TestMethod]
+        public void DeclarationPattern_Discard_DoesNotFail()
+        {
+            const string code = @"
+if (arg is Exception _)
+{
+    Tag(""Arg"", arg);
+}
+Tag(""End"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.DeclarationPattern);
+            validator.ValidateTag("Arg", x => x.Should().BeNull());  // ToDo: MMF-2563 should have NotNull instead
+            validator.TagValues("End").Should().HaveCount(1).And.OnlyContain(x => x == null);
+        }
+
+        [TestMethod]
+        public void DeclarationPattern_Var_PreservePreviousConstraint_DoesNotSetNotNullConstraint()
+        {
+            const string code = @"
+if (arg is var value)
+{
+    Tag(""Value"", value);
+    Tag(""Arg"", arg);
+}
+Tag(""End"", arg);";
+            var setter = new PreProcessTestCheck(OperationKind.ParameterReference, x => x.SetSymbolConstraint(x.Operation.Instance.TrackedSymbol(), TestConstraint.First));
+            var validator = SETestContext.CreateCS(code, ", object arg", setter).Validator;
+            validator.ValidateContainsOperation(OperationKind.DeclarationPattern);
+            validator.ValidateTag("Value", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("Value", x => x.HasConstraint<ObjectConstraint>().Should().BeFalse("'var' only propagates existing constraints"));
+            validator.ValidateTag("Arg", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("Arg", x => x.HasConstraint<ObjectConstraint>().Should().BeFalse("'var' only propagates existing constraints"));
+            validator.TagValues("End").Should().HaveCount(2).And.OnlyContain(x => x != null && x.HasConstraint(TestConstraint.First));     // 2x because value has different states
+        }
+
+        [TestMethod]
+        public void DeclarationPattern_Var_Deconstruction_DoesNotSetConstraint()
```

Comment:
The outermost operation is a recursive operation in both cases.
```suggestion
        public void RecursivePattern_Var_Deconstruction_DoesNotSetConstraint()
```

Reply:
Should go under `RecursivePattern_SetsNotNull_Nested()`

Reply:
We're testing the DeclarationPattern behavior with it. This one is hard to name.